doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -926,14 +909,257 @@ the conversion in hypotheses :n:`{+ @ident}`.
+   to the current section or module, in which case the section and module
+   behavior is the same as :cmd:`Opaque` and :cmd:`Transparent` (without :attr:`global`).
+
+   Levels can be one of the following (higher to lower):
+
+    + ``opaque`` : level of opaque constants. They cannot be expanded by
+      tactics (behaves like +∞, see next item).
+    + :n:`@integer` : levels indexed by an integer. Level 0 corresponds to the
+      default behavior, which corresponds to transparent constants. This
+      level can also be referred to as ``transparent``. Negative levels
+      correspond to constants to be expanded before normal transparent
+      constants, while positive levels correspond to constants to be
+      expanded after normal transparent constants.
+    + ``expand`` : level of constants that should be expanded first (behaves
+      like −∞)
+    + ``transparent`` : Equivalent to level 0
+
+.. cmd:: Print Strategy @reference
+
+   This command prints the strategy currently associated with :n:`@reference`. It
+   fails if :n:`@reference` is not an unfoldable reference, that is, neither a
+   variable nor a constant.
+
+   .. exn:: The reference is not unfoldable.
+      :undocumented:
+
+.. cmd:: Print Strategies
+
+   Print all the currently non-transparent strategies.
+
+.. cmd:: Declare Reduction @ident := @red_expr
+
+   Declares a short name for the reduction expression :n:`@red_expr`, for
+   instance ``lazy beta delta [foo bar]``. This short name can then be used
+   in :n:`Eval @ident in` or ``eval`` constructs. This command
+   accepts the :attr:`local` attribute, which indicates that the reduction
+   will be discarded at the end of the
+   file or module. The name is not qualified. In
+   particular declaring the same name in several modules or in several
+   functor applications will be rejected if these declarations are not
+   local. The name :n:`@ident` cannot be used directly as an Ltac tactic, but
+   nothing prevents the user from also performing a
+   :n:`Ltac @ident := @red_expr`.
+
+   .. seealso:: :ref:`applyingconversionrules`
+
+.. todo PR
+   move rewriting.rst to equality.rst once the text is final,
+   perhaps in a different PR to keep the history somewhat understandable
+
+Equality
+--------
+
+.. tacn:: f_equal
+
+   For a goal with the form :n:`f a__1 ... a__n = g b__1 ... b__n`, creates
+   subgoals :n:`f = g` and :n:`a__i = b__i` for the `n` arguments. SImple subgoals, such as those
+   provable by :tacn:`reflexivity` or :tacn:`congruence` are solved automatically.
+
+.. tacn:: reflexivity
+
+   For a goal with the form `t=u`, verifies that `t` and `u` are :term:`convertible`,
+   and if so,
+   solves the goal.  If not, it fails.  It is equivalent to ``apply refl_equal``.
+
+   .. todo PR: this doesn't cover how reflexibity can solve the goal "True"
+
+      also for te goal "1 + 2 = 4", this generates "Unable to unify ..." with a stack
+      trace.  Should have a better message, right?
```

Comment:
What's wrong with `Unable to unify "4" with "1 + 2".`?  That seems perfectly fine to me.  What error message do you want to see?

Reply:
I don't expect it to refer to unification at all.  A reference to not being convertible would make more sense.

Reply:
What about for the goal `exists y, y + 1 = 4`, running `eexists; reflexivity`?  Unification subsumes conversion, and almost all tactics use unification rather than conversion.  (Essentially the only exception is the `change` family and, I suppose, the reduction tactics.). The difference between unification and conversion is that unification is allowed to instantiate evars in the process of establishing definitional equality, while conversion is not.